src/main/java/pl/sdacademy/carrental/services/ClientService.java

```diff
@@ -0,0 +1,88 @@
+package pl.sdacademy.carrental.services;
+
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import pl.sdacademy.carrental.domain.Address;
+import pl.sdacademy.carrental.domain.Client;
+import pl.sdacademy.carrental.model.ClientForm;
+import pl.sdacademy.carrental.repositories.AddressRepository;
+import pl.sdacademy.carrental.repositories.ClientRepository;
+
+import java.util.List;
+
+@Service
+@Transactional
+public class ClientService {
+
+    private final ClientRepository clientRepository;
+    private final AddressRepository addressRepository;
+
+    public ClientService(ClientRepository clientRepository, AddressRepository addressRepository) {
+        this.clientRepository = clientRepository;
+        this.addressRepository = addressRepository;
+    }
+
+    public List<Client> getAll() {
+        return clientRepository.findAll();
+    }
+
+    public void createClient(final ClientForm clientForm) {
+
+        final Address address = createAddressFromClientForm(clientForm);
+
+        final Client client = Client.builder()
+                .firstName(clientForm.getFirstName())
+                .lastName(clientForm.getLastName())
+                .email(clientForm.getEmail())
+                .address(address)
+                .phoneNumber(clientForm.getPhoneNumber())
+                .build();
+        clientRepository.save(client);
+    }
+
+    public void delete(final Long id) {
+        final Client client = clientRepository.findById(id).orElseThrow();
+        clientRepository.delete(client);
+    }
+
+    public void updateClient(final Long id, final ClientForm clientForm) {
+
+        final Address address = createAddressFromClientForm(clientForm);
+
+        final Client client = clientRepository.findById(id).orElseThrow();
+        client.setFirstName(clientForm.getFirstName());
+        client.setLastName(clientForm.getLastName());
+        client.setAddress(address);
+        clientRepository.save(client);
+    }
+
+    private Address createAddressFromClientForm(final ClientForm clientForm){
+        final Address address = Address.builder()
+                .city(clientForm.getCity())
+                .street(clientForm.getStreet())
+                .building(clientForm.getBuilding())
+                .apartment(clientForm.getApartment())
+                .zipCode(clientForm.getZipCode())
+                .build();
+        return addressRepository.save(address);
+    }
+    // I don't know if saving both branch and client address to the same repository is a good idea
+    // Split, refactor AddressRepository to ClientAddressRepository and BranchAddressRepository?
+    // same with Address entity? What do you think?
```

Comment:
Both respresent same set of data - address, consisting of identical pieces / fields. Having two tables, entity classes and repos serve one and the same purpose contradicts DRY principle.
What if, at some point, we will have to store adresses of employees, or if cars could be delivered to clients (to addresses other than the ones used for billing)? In such cases we would have to store and handle these pieces of data separately as well. In my opinion it is best to keep the addresses table, address entity class and Address Repo as it is - after all a *branch address* is no different from *client billing address* or _prospective_ delivery address in term of its structure.

Reply:
Ouch, didn't delete comment about Adress. I commented that but resolved already during writing.

The question about UserController and AdminSideUserController stays though.